Watersmart 518 python skeleton and sitemap generation

cschroed-usgs opened the pull request:
This is the first rough cut of serving the skeleton pages to searchbots. There is definitely room to improve. Let me know what you think!

cschroed-usgs commented on nwc/src/main/java/gov/usgs/cida/nwc/util/PrettyUglyUrlMapper.java:
```diff
@@ -0,0 +1,124 @@
+import java.nio.charset.Charset;
+import java.util.List;
+import java.util.Locale;
+import org.apache.http.NameValuePair;
```
These two imports are the only reason I brought in org.apache.httpcomponents.httpclient. Is that legit?

jiwalker-usgs replied:
It is probably fine, maybe a bit of overkill, but better than writing it again.

cschroed-usgs replied:
Cool. Thanks!

cschroed-usgs commented on nwc/src/main/java/gov/usgs/cida/nwc/util/PrettyUglyPair.java:
```diff
@@ -0,0 +1,11 @@
+package gov.usgs.cida.nwc.util;
```
This should probably be in the test package -- that is the only place it is used.

cschroed-usgs commented on nwc/src/main/java/gov/usgs/cida/nwc/util/PrettyUglyUrlMapper.java:
```diff
@@ -0,0 +1,124 @@
+				uriBuilder.addParameter(SEARCHBOT_ESCAPED_FRAGMENT_PARAM_NAME, fragment.substring(1));
+				uriBuilder.setFragment(null);
+			} else {
+				throw new IllegalArgumentException(
```
Not sure this contract makes sense -- I throw an exception if there's a fragment that isn't started by a "!", but if there is no fragment, then I pass the same url back to the caller silently. Should I throw exceptions in both cases, or pass the same url back to the caller in both cases?

cschroed-usgs replied:
I switched this to pass the same url back to the caller in both cases.

cschroed-usgs commented on nwc/src/main/webapp/WEB-INF/web.xml:
```diff
@@ -46,17 +46,24 @@
-</web-app>
\ No newline at end of file
+        <url-pattern>/proxywps/*</url-pattern>
+    </servlet-mapping>
```
Apparently there was a re-formatting. My bad. The following is the only thing that matters:

cschroed-usgs commented on nwc/src/test/java/gov/usgs/cida/nwc/util/PrettyUglyUrlMapperTest.java:
```diff
@@ -0,0 +1,92 @@
+
+public class PrettyUglyUrlMapperTest {
+
+	public static final List<PrettyUglyPair> prettyUglyPairs = Arrays.asList(
```
These are slightly modified versions of the example mappings on Google's spec.

cschroed-usgs commented:
I should note that this pull request does not include functionality to serve the sitemap xml files because the files are available by default at urls like <context-path>/sitemap.xml.

jiwalker-usgs commented on nwc/pom.xml:
```diff
@@ -290,11 +294,25 @@
-
+	<dependency>
+		<groupId>org.mockito</groupId>
+		<artifactId>mockito-core</artifactId>
```
Do you want to change this scope to test?  I think you mention in a comment moving stuff to a test package.

cschroed-usgs replied:
Good catch! Will do!

cschroed-usgs commented:
One thing to keep in mind -- I need to generalize this skeleton serving functionality to multiple projects. Some re-design will have to happen there. I'm thinking that the logic for converting a pretty url to a resource name will change from project to project, but the stuff that relies on the Google spec will stay the same.

cschroed-usgs commented:
In other words, the SkeletonPageServlet will have to become more general. Maybe the SkeletonPageServlet's constructor should accept an object that implement a PrettyUrlToResourceName interface. Each project would implement their own and specify it as an init param to SearchBotUglyUrlFilter in the web.xml. The filter would instantiate the parameterized class name and pass the instance to the SkeletonPageServlet's constructor. The servlet would use the PrettyUrlToResourceName instance to perform project-specific resource locating.

jiwalker-usgs commented on nwc/src/main/java/gov/usgs/cida/nwc/servlet/SkeletonPageServlet.java:
```diff
@@ -0,0 +1,88 @@
+	 */
+	String getUrlWithoutContextPath(String fullUrl, String contextPath){
+		if(null == contextPath || contextPath.isEmpty()){
+			throw new IllegalArgumentException("Received empty context path. Context path must be non-empty");
```
I'm not sure whether we can assume that we have a contextPath.  We don't do it much here, but it is a reasonable practice to deploy apps on the ROOT context.  This would cause problems in this case.

cschroed-usgs replied:
Yeah, I was being lazy. Thanks for bringing this up. I will address this.

jiwalker-usgs commented on nwc/src/main/java/gov/usgs/cida/nwc/util/PrettyUglyUrlMapper.java:
```diff
@@ -0,0 +1,124 @@
+ * 
+ */
+public class PrettyUglyUrlMapper {
+	static final String SEARCHBOT_ESCAPED_FRAGMENT_PARAM_NAME = "_escaped_fragment_";
```
This is defined twice, you might be able to find a common place for it.

cschroed-usgs replied:
Done!

jiwalker-usgs commented on nwc/src/test/java/gov/usgs/cida/nwc/util/PrettyUglyUrlMapperTest.java:
```diff
@@ -0,0 +1,92 @@
+		new PrettyUglyPair("/path?queryparams#!hashfragment", "/path?queryparams&_escaped_fragment_=hashfragment")
+	);
+	
+	public PrettyUglyUrlMapperTest() {
```
Just a suggestion, I usually don't include constructor , @Befores, @Afters if they aren't doing anything.

cschroed-usgs replied:
That's a good idea. I will do that.

jiwalker-usgs commented:
It might be useful to look at some old code we had for dealing with peculiarities with URLs.  It is a bit dated, so might not fit with all the weirdness we've introduced since, but it has some golden oldies in there.  https://github.com/USGS-CIDA/tomcat-utilities

cschroed-usgs commented:
Thanks for the suggestion. I'll check those tomcat utils out.

cschroed-usgs commented:
@jiwalker-usgs, I addressed all your comments so fat. The tomcat-utils repo didn't have anything I could use, but it's good to know it's there. Any other comments?

cschroed-usgs commented:
I'm going to defer the generalization refactoring that I rambled about above. Will address in a separate pull request.

jiwalker-usgs commented:
:+1: good to go

cschroed-usgs commented:
Thanks!
